File: gui/src/app/lib/search-parser/scanner.ts
export enum Tag {
  EOF = 'EOF',

  Mention = '@',
  Publication = '~',

  QuotedString = 'QUOTED_STRING',
  Word = 'WORD',
  Regexp = 'REGEXP',

  Whitespace = 'WHITESPACE',
  Error = 'ERROR',
}

export class Tok {

  constructor(
    public tag: Tag,
    public lexeme: string,
    public start: number,
    public end: number,
    public error: string = null
  ) {
  }
}

export function Scan(str: string): Tok[] {
  const scanner: Scanner = new Scanner(str);
  let tokens: Tok[] = [];
  while (true) {
    const tok: Tok = scanner.next();
    tokens.push(tok);
    if (tok.tag === Tag.EOF) {
      break;
    }
  }
  return tokens;
}

export class Scanner {

  input: string[] = [];
  pos: number = 0;
  offset: number = 0;

  constructor(str: string) {
    this.input = (str || '').split('');
  }

  next(): Tok {
    if (this.atEOF()) {
      return this.emit(Tag.EOF);
    }
    const c = this.nextChar();
    switch (c) {
      case '@':
        return this.emit(Tag.Mention);
      case '~':
        return this.emit(Tag.Publication);
      case '"':
        return this.scanQuotedString();
      case '/':
        return this.scanRegexp();
      default:
        if (this.isWhitespace(c)) {
          return this.emit(Tag.Whitespace);
        }
        if (this.isValidInputChar(c)) {
          return this.scanWord();
        }
        return this.emitError(`unknown entity: ${c}`);
    }
  }

  private skipWhitespace() {
    while (!this.atEOF() && this.isWhitespace(this.peekChar())) {
      this.nextChar();
    }
    this.offset = this.pos;
  }

  private isWhitespace(char: string): boolean {
    return char.match(/[\s]/) !== null;
  }

  private scanWord(): Tok {
    while (!this.atEOF() && this.isValidInputChar(this.peekChar()) && !this.isWhitespace(this.peekChar())) {
      this.nextChar();
    }
    return this.emit(Tag.Word);
  }

  private scanQuotedString(): Tok {
    while (!this.matchNextChar('"')) {
      if (this.atEOF()) {
        // implicit quote close on EOF
        return this.emit(Tag.QuotedString);
      }
      this.nextChar();
    }
    return this.emit(Tag.QuotedString);
  }

  private scanRegexp(): Tok {
    while (!this.matchNextChar('/')) {
      if (this.atEOF()) {
        // implicit quote close on EOF
        return this.emit(Tag.Regexp);
      }
      this.nextChar();
    }
    return this.emit(Tag.Regexp);
  }

  private nextChar(): string {
    this.pos++;
    return this.input[this.pos - 1];
  }

  private matchNextChar(c: string): boolean {
    if (this.atEOF() || this.peekChar() != c) {
      return false;
    }
    this.nextChar();
    return true;
  }

  private peekChar(): string {
    return this.input[this.pos];
  }

  private emit(tag: Tag): Tok {
    const lexeme = this.input.slice(this.offset, this.pos).join('');
    const start = this.offset;
    const end = this.pos;
    // advance
    this.offset = this.pos;
    //emit
    return new Tok(tag, lexeme, start, end);
  }

  private emitError(reason: string): Tok {
    const lexeme = this.input.slice(this.offset, this.input.length).join('');
    const start = this.offset;
    const end = this.input.length;
    // advance
    this.offset = this.input.length;
    //emit
    return new Tok(Tag.Error, lexeme, start, end);
  }

  private atEOF(): boolean {
    return this.pos >= this.input.length || this.pos === -1;
  }

  private isValidInputChar(r: string): boolean {
    return r != '@' && r != '~' && r != `"`;
  }
}

File: server/pkg/searchterms/scanner.go
package searchterms

import (
	"fmt"
	"strings"
	"unicode"
)

type tag string

const (
	tagEOF tag = "EOF"

	tagMention     = "@"
	tagPublication = "~"

	tagQuotedString = "QUOTED_STRING"
	tagWord         = "WORD"
)

type token struct {
	tag    tag
	lexeme string
}

func (t token) String() string {
	return fmt.Sprintf("{%s: '%s'}", string(t.tag), t.lexeme)
}

func Scan(str string) ([]token, error) {

	scanner := newScanner(str)

	tokens := make([]token, 0)
	for {
		tok, err := scanner.Next()
		if err != nil {
			return nil, err
		}
		tokens = append(tokens, tok)

		if tok.tag == tagEOF {
			break
		}
	}
	return tokens, nil
}

func newScanner(str string) *scanner {
	return &scanner{
		input:  []rune(str),
		pos:    0,
		offset: 0,
	}
}

type scanner struct {
	input  []rune
	pos    int
	offset int
}

// Next gets the next token, advancing the scanner.
func (s *scanner) Next() (token, error) {
	return s.next()
}

func (s *scanner) next() (token, error) {
	s.skipWhitespace()
	if s.atEOF() {
		return s.emit(tagEOF), nil
	}
	switch r := s.nextRune(); r {
	case '@':
		return s.emit(tagMention), nil
	case '~':
		return s.emit(tagPublication), nil
	case '"':
		return s.scanString()
	default:
		if isValidInputRune(r) {
			return s.scanWord()
		}
		return s.error("unknown entity")
	}
}

func (s *scanner) nextRune() rune {
	r := s.input[s.pos]
	s.pos++
	return r
}

// matchNextRune will match the next rune of a multi-run tag e.g. >= !=
func (s *scanner) matchNextRune(r rune) bool {
	if s.atEOF() || s.peekRune() != r {
		return false
	}
	s.nextRune()
	return true
}

func (s *scanner) skipWhitespace() {
	for !s.atEOF() && unicode.IsSpace(s.peekRune()) {
		s.nextRune()
	}
	s.offset = s.pos
}

func (s *scanner) scanString() (token, error) {
	for !s.matchNextRune('"') {
		if s.atEOF() {
			return s.error("unclosed double quote")
		}
		s.nextRune()
	}
	return trimTokenLexeme(s.emit(tagQuotedString), `""`), nil
}

func (s *scanner) scanWord() (token, error) {
	for !s.atEOF() && isValidInputRune(s.peekRune()) && !isWhitespace(s.peekRune()) {
		s.nextRune()
	}
	return s.emit(tagWord), nil
}

func (s *scanner) atEOF() bool {
	return s.pos >= len(s.input)
}

func (s *scanner) peekRune() rune {
	return s.input[s.pos]
}

func (s *scanner) emit(tag tag) token {
	lexeme := string(s.input[s.offset:s.pos])
	s.offset = s.pos
	return token{tag: tag, lexeme: lexeme}
}

func (s *scanner) error(reason string) (token, error) {
	return token{}, fmt.Errorf("failed to scan string at position %d ('%s'): %s", s.pos, string(s.input[s.offset:s.pos]), reason)
}

func isWhitespace(r rune) bool {
	return r == ' ' || r == '\t' || r == '\n'
}

func isValidInputRune(r rune) bool {
	return r != '@' && r != '~' && r != '"'
}

func trimTokenLexeme(t token, trimSet string) token {
	t.lexeme = strings.Trim(t.lexeme, trimSet)
	return t
}
